### PR TITLE
IB: MIN/MAX trx_id by querying VTQ with commit ts [closes #168]

### LINKED ISSUE #168: Improve vtq_query_commit_ts() algorithm.
`vtq_query_commit_ts()` queries `VTQ` by `commit_ts` Several rows with the same timestamp can exist. And currently function returns first encountered. Instead of that it should return a row with min or max (based on `backwards` argument) `trx_id` to match algorithms from [here](https://github.com/tempesta-tech/mariadb_10.2/wiki/Ordering-of-transactions)

> If there are multiple VTQ records at COMMIT_TS0, then select one corresponding to max(COMMIT_ID).

### PR DESCRIPTION
**Pull request by kevgs**

### CODE REVIEW

File: storage/innobase/vers/vers0vtq.cc [hunk 1]
```diff
@@ -217,6 +225,58 @@ operator< (const timeval &a, const timeval &b)
...
+	btr_pcur_t best;
+	btr_pcur_init(&best);
+	btr_pcur_copy_stored_position(&best, &pcur);
+	trx_id_t best_trx_id = read_trx_id(rec);
+
+	while (true) {
+		if (backwards) {
+			btr_pcur_move_to_prev_user_rec(&pcur, &mtr);
+		} else {
+			btr_pcur_move_to_next_user_rec(&pcur, &mtr);
```
Comment (midenok) [inline storage/innobase/vers/vers0vtq.cc]: @return	TRUE if the cursor moved forward, ending on a user record

File: storage/innobase/vers/vers0vtq.cc [hunk 2]
```diff
@@ -217,6 +225,58 @@ operator< (const timeval &a, const timeval &b)
...
+		}
+
+		if (!btr_pcur_is_on_user_rec(&pcur))
```
Comment (midenok) [inline storage/innobase/vers/vers0vtq.cc]: Needless call.

File: storage/innobase/vers/vers0vtq.cc [hunk 3]
```diff
@@ -217,6 +225,58 @@ operator< (const timeval &a, const timeval &b)
...
+
+		timeval tv;
+		rec = btr_pcur_get_rec(&pcur);
+		rec_get_timeval(rec, 0, tv);
+		if (tv != commit_ts)
+			break;
+
+		trx_id_t trx_id = read_trx_id(rec);
+		if ((backwards && trx_id < best_trx_id) ||
+			(!backwards && trx_id > best_trx_id)) {
```
Comment (midenok) [inline storage/innobase/vers/vers0vtq.cc]: Replace this abracadabra with ternary condition.

**Review by midenok — APPROVED** [review]
Did you test this code? Can we invent some test plan for it?

**Comment by kevgs** [comment]
Same `commit_ts` for several records if possible if `INSERT/UPDATE` several row in one query. But in that case they all will have the same `trx_id`.

It is theoretically possible to have two records with same `commit_ts` but different `trx_id` in case when two threads insert a row each. But I have no idea how to write such a test.

**Comment by midenok** [comment]
I believe it is possible to have multiple clients connected concurrently in one test. You may check for examples in existing tests. I think I saw some.

**Comment by kevgs** [comment]
Sure it's possible. But we need to insert two rows with 0.000001s precision. And test must be stable. 0.000001s even isn't enough for perl test script to switch threads. That looks like a problem.

**Comment by midenok** [comment]
Can you check what is time spread between 2 inserts? I'm pretty curious about it! You may post an example of such test plan here. Just 2 inserts and then select.

**Comment by kevgs** [comment]
Debug build.

```sql
create table t1 (a int) with system versioning engine innodb;
insert into t1 values (1);
insert into t1 values (2);
select * from information_schema.innodb_vtq;
trx_id	commit_id	begin_ts	commit_ts	iso_level
1285	1286	2017-04-05 17:37:10.246418	2017-04-05 17:37:10.246565	RR
1283	1284	2017-04-05 17:37:10.245662	2017-04-05 17:37:10.245843	RR
```

**Comment by midenok** [comment]
Where is multithreading?

**Pull request merged**